File: ios/BlueHotspot/Bluetooth/BleProtocol.swift
import CoreBluetooth
import Foundation

struct BleUuids {
    nonisolated(unsafe) static let service = CBUUID(string: "8B5F0001-7D3E-4E4A-9F1C-4B20F4E9A001")
    nonisolated(unsafe) static let deviceInfo = CBUUID(string: "8B5F0002-7D3E-4E4A-9F1C-4B20F4E9A001")
    nonisolated(unsafe) static let command = CBUUID(string: "8B5F0003-7D3E-4E4A-9F1C-4B20F4E9A001")
    nonisolated(unsafe) static let event = CBUUID(string: "8B5F0004-7D3E-4E4A-9F1C-4B20F4E9A001")
    nonisolated(unsafe) static let pairing = CBUUID(string: "8B5F0005-7D3E-4E4A-9F1C-4B20F4E9A001")
}

enum BleMessage: UInt8 {
    case hello = 1
    case helloAck = 2
    case getStatus = 3
    case status = 4
    case startHotspot = 5
    case hotspotStarting = 6
    case hotspotReady = 7
    case hotspotFailed = 8
    case stopHotspot = 9
    case hotspotStopped = 10
    case ping = 11
    case pong = 12
    case error = 13
}

struct BleFrame {
    let type: BleMessage
    let requestId: UUID
    let payload: String
}

enum BleProtocolError: Error {
    case invalidFrame
    case invalidLength
    case unsupportedVersion
    case invalidUtf8
    case unknownMessage
}

enum BleFrameCodec {
    static func encode(_ frame: BleFrame) throws -> Data {
        guard let payload = frame.payload.data(using: .utf8), payload.count <= 512 else {
            throw BleProtocolError.invalidLength
        }
        var body = Data([1, frame.type.rawValue])
        body.append(contentsOf: withUnsafeBytes(of: frame.requestId.uuid) { Array($0) })
        body.append(payload)
        guard body.count <= 530 else { throw BleProtocolError.invalidLength }
        var result = Data([UInt8(body.count >> 8), UInt8(body.count & 0xff)])
        result.append(body)
        return result
    }
}

struct BleFrameDecoder {
    private var buffer = Data()

    mutating func append(_ fragment: Data) throws -> [BleFrame] {
        buffer.append(fragment)
        guard buffer.count <= 530 * 4 else {
            buffer.removeAll()
            throw BleProtocolError.invalidLength
        }
        var frames: [BleFrame] = []
        while buffer.count >= 2 {
            let bodyLength = (Int(buffer[0]) << 8) | Int(buffer[1])
            guard bodyLength >= 18 && bodyLength <= 530 else {
                buffer.removeAll()
                throw BleProtocolError.invalidLength
            }
            guard buffer.count >= bodyLength + 2 else { break }
            let body = buffer.subdata(in: 2..<(bodyLength + 2))
            buffer.removeSubrange(0..<(bodyLength + 2))
            guard body[0] == 1 else { throw BleProtocolError.unsupportedVersion }
            guard let type = BleMessage(rawValue: body[1]) else { throw BleProtocolError.unknownMessage }
            guard let payload = String(data: body.subdata(in: 18..<body.count), encoding: .utf8) else {
                throw BleProtocolError.invalidUtf8
            }
            let requestId = UUID(uuid: (
                body[2], body[3], body[4], body[5],
                body[6], body[7], body[8], body[9],
                body[10], body[11], body[12], body[13],
                body[14], body[15], body[16], body[17]
            ))
            frames.append(BleFrame(type: type, requestId: requestId, payload: payload))
        }
        return frames
    }
}

File: android/server/src/main/java/io/github/mouse233/bluehotspot/server/ble/BleProtocol.kt
package io.github.mouse233.bluehotspot.server.ble

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.CodingErrorAction
import java.util.UUID

private const val PROTOCOL_VERSION = 1
private const val HEADER_SIZE = 18
private const val MAX_PAYLOAD_SIZE = 512
private const val MAX_BODY_SIZE = HEADER_SIZE + MAX_PAYLOAD_SIZE

internal enum class BleMessage(val code: Int) {
    HELLO(1), HELLO_ACK(2), GET_STATUS(3), STATUS(4), START_HOTSPOT(5),
    HOTSPOT_STARTING(6), HOTSPOT_READY(7), HOTSPOT_FAILED(8), STOP_HOTSPOT(9),
    HOTSPOT_STOPPED(10), PING(11), PONG(12), ERROR(13);

    companion object {
        fun fromCode(code: Int): BleMessage = entries.firstOrNull { it.code == code }
            ?: throw BleProtocolException("unknown message type: $code")
    }
}

internal data class BleFrame(
    val type: BleMessage,
    val requestId: UUID,
    val payload: String = "",
)

internal class BleProtocolException(message: String) : Exception(message)

internal object BleFrameCodec {
    fun encode(frame: BleFrame): ByteArray {
        val payload = frame.payload.toByteArray(Charsets.UTF_8)
        require(payload.size <= MAX_PAYLOAD_SIZE) { "payload is too large" }
        val body = ByteBuffer.allocate(HEADER_SIZE + payload.size).order(ByteOrder.BIG_ENDIAN)
            .put(PROTOCOL_VERSION.toByte())
            .put(frame.type.code.toByte())
            .putLong(frame.requestId.mostSignificantBits)
            .putLong(frame.requestId.leastSignificantBits)
            .put(payload)
            .array()
        return ByteBuffer.allocate(body.size + 2).order(ByteOrder.BIG_ENDIAN)
            .putShort(body.size.toShort())
            .put(body)
            .array()
    }

    fun decode(body: ByteArray): BleFrame {
        if (body.size !in HEADER_SIZE..MAX_BODY_SIZE) {
            throw BleProtocolException("invalid body length: ${body.size}")
        }
        val buffer = ByteBuffer.wrap(body).order(ByteOrder.BIG_ENDIAN)
        if (buffer.get().toInt() != PROTOCOL_VERSION) {
            throw BleProtocolException("unsupported protocol version")
        }
        val type = BleMessage.fromCode(buffer.get().toInt() and 0xff)
        val requestId = UUID(buffer.long, buffer.long)
        val payloadBytes = ByteArray(buffer.remaining())
        buffer.get(payloadBytes)
        val payload = try {
            Charsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
                .decode(ByteBuffer.wrap(payloadBytes))
                .toString()
        } catch (error: Exception) {
            throw BleProtocolException("payload is not valid UTF-8: ${error.message}")
        }
        return BleFrame(type, requestId, payload)
    }
}

/** Reassembles frames when a characteristic write contains only a fragment. */
internal class BleFrameDecoder {
    private var buffer = ByteArray(0)

    fun append(fragment: ByteArray): List<BleFrame> {
        if (fragment.isEmpty()) return emptyList()
        if (buffer.size + fragment.size > MAX_BODY_SIZE * 4) {
            buffer = ByteArray(0)
            throw BleProtocolException("frame stream is too large")
        }
        buffer += fragment
        val frames = mutableListOf<BleFrame>()
        while (buffer.size >= 2) {
            val bodyLength = ((buffer[0].toInt() and 0xff) shl 8) or
                (buffer[1].toInt() and 0xff)
            if (bodyLength !in HEADER_SIZE..MAX_BODY_SIZE) {
                buffer = ByteArray(0)
                throw BleProtocolException("invalid frame length: $bodyLength")
            }
            if (buffer.size < bodyLength + 2) break
            val body = buffer.copyOfRange(2, bodyLength + 2)
            buffer = buffer.copyOfRange(bodyLength + 2, buffer.size)
            frames += BleFrameCodec.decode(body)
        }
        return frames
    }
}

